File: fruitfly/brain.py
"""Leaky integrate-and-fire simulation of the whole FlyWire connectome.

Model follows Shiu et al. 2024 ("A Drosophila computational brain model
reveals sensorimotor processing", Nature 634:210-219):

  - every classified neuron is one LIF unit
  - connection weight = summed synapse count between the pair
  - sign from the predicted neurotransmitter (ACh +, GABA/Glu -)
  - membrane time constant 20 ms, resting/reset -52 mV, threshold -45 mV,
    refractory 2.2 ms, synaptic time constant 5.5 ms, delay 1.8 ms,
    single-synapse PSP calibrated to 0.275 mV

Differences from the paper (documented, deliberate, for real-time use):
  - exponential (not alpha) synaptic kernel, calibrated to the same peak PSP
  - default dt = 2.0 ms instead of 0.1 ms (decay factors are exact,
    exp(-dt/tau), so dt costs spike-timing resolution but does not
    rescale the time constants themselves)
  - optional Poisson background noise so the brain has spontaneous activity
    (the paper's network is silent without stimulation; a real brain is not)
  - spike-frequency adaptation (biological, and it stabilizes the network:
    without it the recurrent excitation has a hard phase transition between
    total silence and runaway firing; with it, supercritical noise produces
    intermittent self-quenching bursts — the fly's spontaneous "moods")

The propagation is event-driven: per step only the outgoing rows of neurons
that actually spiked are touched, which keeps ~50M synapses tractable on a
CPU because fly brain activity is sparse.
"""

from __future__ import annotations

import math

import numpy as np


class Params:
    v_rest = -52.0      # mV
    v_reset = -52.0     # mV
    v_thresh = -45.0    # mV
    tau_m = 20.0        # ms, membrane time constant
    tau_syn = 5.5       # ms, synaptic time constant
    t_refract = 2.2     # ms
    delay = 1.8         # ms, synaptic delay
    psp_peak = 0.275    # mV, depolarization from one synapse at rest
    adapt_b = 8.0       # mV of adaptation added per spike
    tau_adapt = 500.0   # ms, adaptation decay


def _decays(dt: float, p: Params) -> tuple[float, float, float]:
    """Per-step decay factors (synapse, membrane, adaptation).

    exp(-dt/tau), not the forward-Euler (1 - dt/tau) this used to use.
    Euler does not merely add error, it silently rescales the model: at
    dt=2ms it turns tau_syn=5.5 into an effective 4.42ms (-19.5%) and
    tau_m=20 into 18.98 (-5.1%). The exact factor samples the true
    continuous exponential at any dt, so dt no longer distorts the
    kernel and only limits spike-timing resolution. It costs nothing —
    these are precomputed constants either way.
    """
    return (math.exp(-dt / p.tau_syn), math.exp(-dt / p.tau_m),
            math.exp(-dt / p.tau_adapt))


def _psp_calibration(dt: float, p: Params) -> float:
    """Weight w such that one unit-weight synaptic event peaks at psp_peak mV.

    Simulates the linear membrane response to a single synaptic impulse
    with the same integration scheme and the same decay factors used at
    runtime, then rescales. Both must stay in step: this shares _decays
    with `step` precisely so the two cannot drift apart.
    """
    decay_s, decay_m, _ = _decays(dt, p)
    s, v, peak = 1.0, 0.0, 0.0
    for _ in range(int(200 / dt)):
        v = s + (v - s) * decay_m        # v measured from rest; v_inf = s
        s *= decay_s
        peak = max(peak, v)
    return p.psp_peak / peak


class Brain:
    def __init__(self, indptr, indices, weights, pops,
                 dt: float = 2.0,
                 noise_rate: float = 0.0, noise_weight: float = 0.0,
                 exc_gain: float = 1.0, inh_gain: float = 1.5,
                 noise_pop: str | None = "central",
                 lamina_bias: float = 0.0,
                 seed: int | None = None):
        self.p = Params()
        self.dt = float(dt)
        self.n = len(indptr) - 1
        self.indptr = np.ascontiguousarray(indptr, dtype=np.int64)
        self.indices = np.ascontiguousarray(indices, dtype=np.int32)
        self.pops = pops
        self.rng = np.random.default_rng(seed)

        w_unit = _psp_calibration(self.dt, self.p)
        w = np.array(weights, dtype=np.float32) * w_unit
        w[w > 0] *= exc_gain
        w[w < 0] *= inh_gain
        self.weights = np.ascontiguousarray(w)

        self.v = np.full(self.n, self.p.v_rest, dtype=np.float32)
        self.vth = np.full(self.n, self.p.v_thresh, dtype=np.float32)
        self.s = np.zeros(self.n, dtype=np.float32)   # synaptic drive (mV)
        self.a = np.zeros(self.n, dtype=np.float32)   # adaptation (mV)
        d_s, d_m, d_a = _decays(self.dt, self.p)
        self.decay_s = np.float32(d_s)
        self.decay_m = np.float32(d_m)
        self.decay_a = np.float32(d_a)
        self.refract = np.zeros(self.n, dtype=np.int32)
        self.refract_steps = max(1, int(round(self.p.t_refract / self.dt)))

        # synaptic delay ring buffer
        self.delay_steps = max(1, int(round(self.p.delay / self.dt)))
        self.ring = np.zeros((self.delay_steps, self.n), dtype=np.float32)
        self.ring_pos = 0

        # background noise (spontaneity): Poisson synaptic bombardment,
        # regulated by a slow homeostat toward target_rate. Call it what
        # it is: a stability governor on an invented noise floor, not
        # "arousal". It raises the floor when the network goes quiet and
        # backs off when it rages, which is what keeps the network out of
        # the coma/seizure pair it otherwise collapses into. Real arousal
        # is neuromodulatory and largely octopaminergic, and this is not
        # that — octopamine multiplies drive that already exists, so it
        # could never do this job: multiplying a silent network keeps it
        # silent. All actual computation still runs through real synapses.
        # Noise goes only to the central brain by default: the optic lobes
        # (~100k of the 139k neurons) are driven by actual visual input, and
        # noise there makes the fly hallucinate looming objects.
        self.noise_rate = float(noise_rate)          # Hz per neuron (current)
        self.noise_base = float(noise_rate)
        self.noise_weight = np.float32(noise_weight * w_unit)  # in synapses
        if noise_pop and noise_pop in pops and len(pops[noise_pop]):
            self.noise_targets = np.asarray(pops[noise_pop], dtype=np.int64)
        else:
            self.noise_targets = np.arange(self.n, dtype=np.int64)
        self.target_rate = 1.0                       # Hz/neuron network mean
        self._rate_ema = 0.0                         # smoothed Hz/neuron

        # external drive: map neuron index -> forced firing rate (Hz)
        self._stim_idx = np.empty(0, dtype=np.int32)
        self._stim_p = np.empty(0, dtype=np.float32)

        # The giant fiber is one of the largest neurons in the fly and is
        # known for its high firing threshold — escape only triggers on
        # strong, synchronized looming drive. The paper's uniform threshold
        # makes it chronically excitable in a spontaneously active network,
        # so we give it (only it) a higher one.
        self.shift_threshold("GF", +10.0)

        # Lamina monopolar cells are graded, non-spiking neurons held at a
        # depolarized operating point; photoreceptor histamine inhibition
        # is released by darkness (the OFF response). A tonic bias current
        # emulates that operating point in the LIF approximation.
        self.bias = np.zeros(self.n, dtype=np.float32)
        if "lamina" in pops and len(pops["lamina"]):
            self.bias[pops["lamina"]] = np.float32(lamina_bias)

        self.t = 0.0            # simulated ms
        self.total_spikes = 0

    def shift_threshold(self, pop: str, delta_mv: float) -> None:
        if pop in self.pops and len(self.pops[pop]):
            self.vth[self.pops[pop]] += delta_mv

    def reset_state(self) -> None:
        """Fresh dynamical state (same anatomy): a new fly's brain."""
        self.v[:] = self.p.v_rest
        self.s[:] = 0.0
        self.a[:] = 0.0
        self.refract[:] = 0
        self.ring[:] = 0.0
        self.noise_rate = self.noise_base
        self._rate_ema = 0.0
        self._stim_idx = np.empty(0, dtype=np.int32)
        self._stim_p = np.empty(0, dtype=np.float32)

    # ---------------------------------------------------------------- stim
    def set_stimulus(self, stim: dict[str, float] | list):
        """Set forced firing rates for sensory neurons.

        Accepts {population_name: rate_hz} or [(indices_or_name, rates), ...]
        where rates is a scalar Hz or a per-neuron array of Hz (same length
        as the index array — this is how the retina drives each
        photoreceptor at its own rate). Stimulated neurons fire as Poisson
        processes, like the optogenetic-style activation in Shiu et al.
        """
        items = (stim.items() if isinstance(stim, dict) else stim)
        idx_parts, p_parts = [], []
        for key, rate in items:
            idx = self.pops[key] if isinstance(key, str) else key
            if len(idx) == 0:
                continue
            if isinstance(rate, np.ndarray):
                keep = rate > 0.0
                if not keep.any():
                    continue
                idx_parts.append(np.asarray(idx, dtype=np.int32)[keep])
                p_parts.append((rate[keep] * self.dt * 1e-3)
                               .astype(np.float32))
                continue
            if rate <= 0.0:
                continue
            idx_parts.append(np.asarray(idx, dtype=np.int32))
            p_parts.append(np.full(len(idx), rate * self.dt * 1e-3,
                                   dtype=np.float32))
        if idx_parts:
            self._stim_idx = np.concatenate(idx_parts)
            self._stim_p = np.concatenate(p_parts)
        else:
            self._stim_idx = np.empty(0, dtype=np.int32)
            self._stim_p = np.empty(0, dtype=np.float32)

    # ---------------------------------------------------------------- step
    def step(self) -> np.ndarray:
        """Advance one dt. Returns indices of neurons that spiked."""
        p, dt = self.p, self.dt

        # deliver delayed synaptic input, decay drive
        self.s *= self.decay_s
        self.s += self.ring[self.ring_pos]
        self.ring[self.ring_pos] = 0.0

        # background noise: sample the total Poisson event count, then
        # scatter events onto random neurons (much cheaper than per-neuron)
        if self.noise_rate > 0.0 and self.noise_weight > 0.0:
            m = len(self.noise_targets)
            lam = m * self.noise_rate * dt * 1e-3
            k = self.rng.poisson(lam)
            if k:
                hits = self.noise_targets[self.rng.integers(0, m, size=k)]
                np.add.at(self.s, hits, self.noise_weight)

        # membrane integration with spike-frequency adaptation.
        # Integrate everyone unconditionally (fast fused array ops), then
        # clamp the (few) refractory neurons back to reset.
        # exponential Euler: exact for input held constant over the step,
        # and unconditionally stable, unlike the forward Euler it replaces
        self.a *= self.decay_a
        v_inf = p.v_rest + self.s - self.a + self.bias
        self.v = v_inf + (self.v - v_inf) * self.decay_m
        ref = np.flatnonzero(self.refract > 0)
        if len(ref):
            self.v[ref] = p.v_reset
            self.refract[ref] -= 1

        # threshold crossing
        spiked = np.flatnonzero(self.v >= self.vth)

        # forced sensory spikes (Poisson at the commanded rate)
        if len(self._stim_idx):
            fire = self.rng.random(len(self._stim_idx),
                                   dtype=np.float32) < self._stim_p
            forced = self._stim_idx[fire]
            forced = forced[self.refract[forced] <= 0]
            if len(forced):
                spiked = np.union1d(spiked, forced)

        if len(spiked):
            self.v[spiked] = p.v_reset
            self.refract[spiked] = self.refract_steps
            self.a[spiked] += p.adapt_b
            self._propagate(spiked)
            self.total_spikes += len(spiked)

        # noise-floor governor: nudge the floor toward the target rate
        inst = len(spiked) * 1000.0 / (self.n * dt)      # Hz/neuron this step
        self._rate_ema += (inst - self._rate_ema) * (dt / 500.0)
        if self.noise_base > 0.0:
            err = self.target_rate - self._rate_ema
            self.noise_rate += err * (dt / 200.0) * self.noise_base * 0.1
            self.noise_rate = min(max(self.noise_rate, 0.0),
                                  self.noise_base * 3.0)

        self.ring_pos = (self.ring_pos + 1) % self.delay_steps
        self.t += dt
        return spiked

    def _propagate(self, spiked: np.ndarray) -> None:
        """Add outgoing weights of spiking neurons into the delay buffer."""
        starts = self.indptr[spiked]
        counts = (self.indptr[spiked + 1] - starts).astype(np.int64)
        total = int(counts.sum())
        if total == 0:
            return
        # flat positions of all outgoing synapses of all spiking neurons
        offsets = np.repeat(starts - np.concatenate(
            ([0], np.cumsum(counts)[:-1])), counts)
        flat = offsets + np.arange(total, dtype=np.int64)
        targets = self.indices[flat]
        contrib = np.bincount(targets, weights=self.weights[flat],
                              minlength=self.n).astype(np.float32)
        slot = (self.ring_pos + self.delay_steps - 1) % self.delay_steps
        self.ring[slot] += contrib

    # ------------------------------------------------------------- helpers
    def pop_count(self, spiked: np.ndarray, name: str) -> int:
        """How many spikes in `spiked` belong to population `name`."""
        return int(np.isin(spiked, self.pops[name], assume_unique=False).sum())


class RateMonitor:
    """Exponential moving average firing rate (Hz/neuron) per population."""

    def __init__(self, brain: Brain, names: list[str], tau_ms: float = 100.0):
        self.brain = brain
        self.names = [n for n in names if len(brain.pops.get(n, [])) > 0]
        self.sizes = {n: len(brain.pops[n]) for n in self.names}
        self.masks = {}
        for n in self.names:
            m = np.zeros(brain.n, dtype=bool)
            m[brain.pops[n]] = True
            self.masks[n] = m
        self.rates = dict.fromkeys(self.names, 0.0)
        self.decay = np.exp(-brain.dt / tau_ms)
        self.gain = (1.0 - self.decay) * 1000.0 / brain.dt  # -> Hz

    def update(self, spiked: np.ndarray) -> None:
        for n in self.names:
            k = int(self.masks[n][spiked].sum()) if len(spiked) else 0
            self.rates[n] = (self.rates[n] * self.decay
                             + self.gain * k / self.sizes[n])

File: web/src/brain/brain.ts
/**
 * Leaky integrate-and-fire simulation of the whole FlyWire connectome —
 * the browser's copy of `fruitfly/brain.py`, ported line for line.
 *
 * The model and its deliberate departures from Shiu et al. 2024 are
 * documented in the Python file and not repeated here; what this file
 * documents is where the *port* had to make a choice, because that is
 * where the two can drift apart.
 *
 * The propagation is event-driven: per step only the outgoing rows of
 * neurons that actually spiked are touched, which is what keeps 2.7 M
 * connections tractable at 60 Hz. Where numpy expressed that with
 * gather/scatter gymnastics to stay out of the interpreter, this is a
 * plain nested loop, which in a JIT is the faster of the two and much
 * the clearer.
 *
 * Everything the model holds per neuron lives in a Float32Array, so the
 * state has the same precision as numpy's. Arithmetic still happens in
 * float64 and is rounded on store, which numpy does not do — one of
 * several reasons parity here is statistical.
 */

import { PARAMS, decays, pspCalibration, pyRound } from "./params.js";
import type { BrainData } from "./loader.js";
import { Rng } from "./rng.js";

export interface BrainOptions {
  dt?: number;
  noiseRate?: number;
  noiseWeight?: number;
  excGain?: number;
  inhGain?: number;
  /** Population the noise floor is delivered to; null spreads it everywhere. */
  noisePop?: string | null;
  seed?: number;
}

/** One stimulus channel: neurons, and the Hz to drive each of them at. */
export type Stimulus =
  | { pop: string; rate: number }
  | { idx: Int32Array | number[]; rate: number | Float32Array | number[] };

export class Brain {
  readonly n: number;
  readonly dt: number;
  readonly pops: Map<string, Int32Array>;

  private readonly indptr: Int32Array;
  private readonly indices: Int32Array;
  /** Signed, PSP-calibrated, gain-applied. The file's int16 stays untouched. */
  private readonly weights: Float32Array;

  private readonly v: Float32Array;
  private readonly vth: Float32Array;
  private readonly s: Float32Array; // synaptic drive, mV
  private readonly a: Float32Array; // adaptation, mV
  private readonly bias: Float32Array; // per-neuron tonic current, mV
  private readonly refract: Int32Array;
  private readonly refractSteps: number;

  private readonly decayS: number;
  private readonly decayM: number;
  private readonly decayA: number;

  // Synaptic delay ring buffer: delaySteps rows of n.
  private readonly ring: Float32Array;
  private readonly delaySteps: number;
  private ringPos = 0;

  private readonly rng: Rng;
  private noiseRate: number;
  private readonly noiseBase: number;
  private readonly noiseWeight: number;
  private readonly noiseTargets: Int32Array;
  private readonly targetRate = 1.0; // Hz/neuron, network mean
  private rateEma = 0.0;

  // Forced sensory drive: parallel arrays of neuron index and per-step
  // spike probability.
  private stimIdx = new Int32Array(0);
  private stimP = new Float32Array(0);

  /**
   * Scratch reused every step. `spiked` is written and read within one
   * step and never escapes, so allocating it per step would hand the GC
   * 60 arrays a second for nothing.
   */
  private readonly spikedBuf: Int32Array;
  private spikedCount = 0;
  /** Marks neurons already counted this step, so a forced spike that also
   *  crossed threshold is one spike and not two. numpy got this from
   *  `union1d` returning sorted unique; here it needs saying. */
  private readonly spikedMark: Uint8Array;

  t = 0.0; // simulated ms
  totalSpikes = 0;

  constructor(data: BrainData, opts: BrainOptions = {}) {
    const p = PARAMS;
    const {
      dt = 2.0,
      noiseRate = 0.0,
      noiseWeight = 0.0,
      excGain = 1.0,
      inhGain = 1.5,
      noisePop = "central",
      seed = 1,
    } = opts;

    this.dt = dt;
    this.n = data.header.n_neurons;
    this.indptr = data.indptr;
    this.indices = data.indices;
    this.pops = data.pops;
    this.rng = new Rng(seed);

    const wUnit = pspCalibration(dt, p);
    this.weights = new Float32Array(data.weights.length);
    for (let i = 0; i < data.weights.length; i++) {
      const w = data.weights[i]! * wUnit;
      this.weights[i] = w > 0 ? w * excGain : w * inhGain;
    }

    this.v = new Float32Array(this.n).fill(p.vRest);
    this.vth = new Float32Array(this.n).fill(p.vThresh);
    this.s = new Float32Array(this.n);
    this.a = new Float32Array(this.n);
    this.bias = new Float32Array(this.n);
    this.refract = new Int32Array(this.n);
    this.refractSteps = Math.max(1, pyRound(p.tRefract / dt));

    const d = decays(dt, p);
    this.decayS = d.s;
    this.decayM = d.m;
    this.decayA = d.a;

    this.delaySteps = Math.max(1, pyRound(p.delay / dt));
    this.ring = new Float32Array(this.delaySteps * this.n);

    this.noiseRate = noiseRate;
    this.noiseBase = noiseRate;
    this.noiseWeight = noiseWeight * wUnit;
    const targets = noisePop === null ? undefined : data.pops.get(noisePop);
    if (targets && targets.length > 0) {
      this.noiseTargets = targets;
    } else {
      // Noise everywhere is the fallback, not the intent: the optic
      // lobes are ~100k of the 139k neurons and are meant to be driven
      // by actual light. Noise in there makes the fly hallucinate
      // looming objects.
      this.noiseTargets = new Int32Array(this.n);
      for (let i = 0; i < this.n; i++) this.noiseTargets[i] = i;
    }

    this.spikedBuf = new Int32Array(this.n);
    this.spikedMark = new Uint8Array(this.n);

    // Constructor order matters and mirrors Python's: vth is filled
    // uniformly, and only then does the giant fiber get its higher
    // threshold. It is one of the largest neurons in the fly and is
    // known for exactly that; the paper's uniform threshold leaves it
    // chronically excitable in a spontaneously active network.
    this.shiftThreshold("GF", 10.0);
  }

  /**
   * Where the noise governor currently sits, in Hz per neuron.
   *
   * Worth reading rather than assuming: it can only *raise* the floor,
   * and this network is already livelier than the 1.0 Hz/neuron it aims
   * at, so it empties to zero within about ten simulated seconds and
   * stays there. The parity gate compares it, because two ports whose
   * rates agreed while their governors sat in different places would be
   * agreeing by luck.
   */
  get noiseFloor(): number {
    return this.noiseRate;
  }

  shiftThreshold(pop: string, deltaMv: number): void {
    const idx = this.pops.get(pop);
    if (!idx) return;
    for (let i = 0; i < idx.length; i++) {
      const j = idx[i]!;
      this.vth[j] = this.vth[j]! + deltaMv;
    }
  }

  /** Fresh dynamical state, same anatomy: a new fly's brain. */
  resetState(): void {
    this.v.fill(PARAMS.vRest);
    this.s.fill(0);
    this.a.fill(0);
    this.refract.fill(0);
    this.ring.fill(0);
    this.noiseRate = this.noiseBase;
    this.rateEma = 0;
    this.stimIdx = new Int32Array(0);
    this.stimP = new Float32Array(0);
  }

  /**
   * Forced firing rates for sensory neurons, as Poisson processes — the
   * optogenetic-style activation of Shiu et al. A per-neuron rate array
   * is how the retina drives every photoreceptor at its own rate.
   */
  setStimulus(stim: Stimulus[]): void {
    const idxParts: number[] = [];
    const pParts: number[] = [];
    const scale = this.dt * 1e-3;
    for (const item of stim) {
      const idx = "pop" in item ? this.pops.get(item.pop) : item.idx;
      if (!idx || idx.length === 0) continue;
      const rate = item.rate;
      if (typeof rate === "number") {
        if (rate <= 0) continue;
        const p = rate * scale;
        for (let i = 0; i < idx.length; i++) {
          idxParts.push(idx[i]!);
          pParts.push(p);
        }
      } else {
        for (let i = 0; i < idx.length; i++) {
          const r = rate[i]!;
          if (r > 0) {
            idxParts.push(idx[i]!);
            pParts.push(r * scale);
          }
        }
      }
    }
    this.stimIdx = Int32Array.from(idxParts);
    this.stimP = Float32Array.from(pParts);
  }

  /**
   * Advance one dt. Returns a view of the neurons that spiked, valid
   * until the next call — copy it if you need to keep it.
   */
  step(): Int32Array {
    const p = PARAMS;
    const { n, dt, v, s, a, vth, bias, refract, ring } = this;
    const ringBase = this.ringPos * n;

    // Deliver the synaptic input that left its presynaptic neuron
    // delay ms ago, and decay what is already in flight.
    for (let i = 0; i < n; i++) {
      s[i] = s[i]! * this.decayS + ring[ringBase + i]!;
      ring[ringBase + i] = 0;
    }

    // Background noise: draw the total event count once and scatter it,
    // which is far cheaper than a draw per neuron. Straight into `s`,
    // with no intermediate array — at ~10k events a step, an allocation
    // here would be the whole cost of the step.
    if (this.noiseRate > 0 && this.noiseWeight > 0) {
      const m = this.noiseTargets.length;
      const lam = m * this.noiseRate * dt * 1e-3;
      const k = this.rng.poisson(lam);
      for (let j = 0; j < k; j++) {
        const i = this.noiseTargets[this.rng.below(m)]!;
        s[i] = s[i]! + this.noiseWeight;
      }
    }

    // Membrane integration with spike-frequency adaptation. Exponential
    // Euler: exact for input held constant across the step, and
    // unconditionally stable. Everyone is integrated and the few
    // refractory neurons are clamped back afterwards, as in Python.
    this.spikedCount = 0;
    for (let i = 0; i < n; i++) {
      a[i] = a[i]! * this.decayA;
      const vInf = p.vRest + s[i]! - a[i]! + bias[i]!;
      let vi = vInf + (v[i]! - vInf) * this.decayM;
      const r = refract[i]!;
      if (r > 0) {
        vi = p.vReset;
        refract[i] = r - 1;
      }
      v[i] = vi;
      if (vi >= vth[i]!) {
        this.spikedBuf[this.spikedCount++] = i;
        this.spikedMark[i] = 1;
      }
    }

    // Forced sensory spikes, Poisson at the commanded rate. A neuron in
    // its refractory period cannot be forced, and one that already
    // crossed threshold this step must not be counted twice.
    for (let j = 0; j < this.stimIdx.length; j++) {
      if (this.rng.next() < this.stimP[j]!) {
        const i = this.stimIdx[j]!;
        if (refract[i]! <= 0 && this.spikedMark[i] === 0) {
          this.spikedBuf[this.spikedCount++] = i;
          this.spikedMark[i] = 1;
        }
      }
    }

    const count = this.spikedCount;
    if (count > 0) {
      const slot = ((this.ringPos + this.delaySteps - 1) % this.delaySteps) * n;
      for (let j = 0; j < count; j++) {
        const i = this.spikedBuf[j]!;
        v[i] = p.vReset;
        refract[i] = this.refractSteps;
        a[i] = a[i]! + p.adaptB;
        this.spikedMark[i] = 0;
        const end = this.indptr[i + 1]!;
        for (let e = this.indptr[i]!; e < end; e++) {
          const target = slot + this.indices[e]!;
          ring[target] = ring[target]! + this.weights[e]!;
        }
      }
      this.totalSpikes += count;
    }

    // Noise-floor governor. A stability governor on an invented noise
    // floor, not arousal: it raises the floor when the network goes
    // quiet and backs off when it rages, which is what keeps it out of
    // the coma/seizure pair it otherwise collapses into.
    const inst = (count * 1000.0) / (n * dt);
    this.rateEma += (inst - this.rateEma) * (dt / 500.0);
    if (this.noiseBase > 0) {
      const err = this.targetRate - this.rateEma;
      this.noiseRate += err * (dt / 200.0) * this.noiseBase * 0.1;
      this.noiseRate = Math.min(
        Math.max(this.noiseRate, 0),
        this.noiseBase * 3.0,
      );
    }

    this.ringPos = (this.ringPos + 1) % this.delaySteps;
    this.t += dt;
    return this.spikedBuf.subarray(0, count);
  }
}

/** Exponential moving average firing rate (Hz/neuron) per population. */
export class RateMonitor {
  readonly names: string[];
  private readonly sizes: Float64Array;
  private readonly rate: Float64Array;
  private readonly hits: Float64Array;
  private readonly decay: number;
  private readonly gain: number;
  /** Neuron -> which watched populations it belongs to, CSR-style. */
  private readonly memberOffsets: Int32Array;
  private readonly memberSlots: Int32Array;

  constructor(brain: Brain, names: string[], tauMs = 100.0) {
    this.names = names.filter((n) => (brain.pops.get(n)?.length ?? 0) > 0);
    this.decay = Math.exp(-brain.dt / tauMs);
    this.gain = ((1.0 - this.decay) * 1000.0) / brain.dt;
    this.sizes = new Float64Array(this.names.length);
    this.rate = new Float64Array(this.names.length);
    this.hits = new Float64Array(this.names.length);

    // Python asks "is this spike in that population" with one boolean
    // mask per population, which costs populations x spikes per step.
    // Inverted into a per-neuron membership table it costs just spikes,
    // and a step can spike thousands of times. A neuron needs a list
    // rather than a slot because the populations overlap — GF sits
    // inside descending, DNa02_L inside DNa02.
    const counts = new Int32Array(brain.n + 1);
    this.names.forEach((name, slot) => {
      const idx = brain.pops.get(name)!;
      this.sizes[slot] = idx.length;
      for (let i = 0; i < idx.length; i++) counts[idx[i]!]! += 1;
    });
    const offsets = new Int32Array(brain.n + 1);
    let total = 0;
    for (let i = 0; i < brain.n; i++) {
      offsets[i] = total;
      total += counts[i]!;
    }
    offsets[brain.n] = total;
    const slots = new Int32Array(total);
    const at = offsets.slice(0, brain.n);
    this.names.forEach((name, slot) => {
      const idx = brain.pops.get(name)!;
      for (let i = 0; i < idx.length; i++) slots[at[idx[i]!]!++] = slot;
    });
    this.memberOffsets = offsets;
    this.memberSlots = slots;
  }

  update(spiked: Int32Array): void {
    this.hits.fill(0);
    for (let j = 0; j < spiked.length; j++) {
      const i = spiked[j]!;
      const end = this.memberOffsets[i + 1]!;
      for (let e = this.memberOffsets[i]!; e < end; e++) {
        this.hits[this.memberSlots[e]!]! += 1;
      }
    }
    for (let slot = 0; slot < this.rate.length; slot++) {
      this.rate[slot] =
        this.rate[slot]! * this.decay +
        (this.gain * this.hits[slot]!) / this.sizes[slot]!;
    }
  }

  get(name: string): number {
    const slot = this.names.indexOf(name);
    return slot < 0 ? 0 : this.rate[slot]!;
  }

  /** All watched rates, for callers that want to post them somewhere. */
  snapshot(): Record<string, number> {
    const out: Record<string, number> = {};
    this.names.forEach((name, slot) => {
      out[name] = this.rate[slot]!;
    });
    return out;
  }
}
